### PR TITLE
chore: add awakeable periodic service

### PR DESCRIPTION
**Pull request by P403n1x87**
## Description

This change adds a variant of the periodic service class that can also be awakened on demand to perform the scheduled work immediately.

A motivation for implementing such behaviour is for when a queue is full and needs to be flushed ahead of schedule to make room for more data.

## Checklist
- [x] Title must conform to [conventional commit](https://github.com/conventional-changelog/commitlint/tree/master/%40commitlint/config-conventional).
- [x] Add additional sections for `feat` and `fix` pull requests.
- [x] Ensure tests are passing for affected code.
- [x] [Library documentation](https://github.com/DataDog/dd-trace-py/tree/1.x/docs) and/or [Datadog's documentation site](https://github.com/DataDog/documentation/) is updated. Link to doc PR in description.

## Reviewer Checklist
- [ ] Title is accurate.
- [ ] Description motivates each change.
- [ ] No unnecessary changes were introduced in this PR.
- [ ] PR cannot be broken up into smaller PRs.
- [ ] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [x] Release note has been added for fixes and features, or else `changelog/no-changelog` label added.
- [ ] All relevant GitHub issues are correctly linked.
- [ ] Backports are identified and tagged with Mergifyio.
- [ ] Add to milestone.

### CODE REVIEW

**Review by ZStriker19 — DISMISSED** [review]
Interested to see how it's used!

File: ddtrace/internal/periodic.py [hunk 1]
```diff
@@ -242,3 +242,141 @@ def on_shutdown():
...
+                break
+
+            self._target()
+
+        if self._on_shutdown is not None:
+            self._on_shutdown()
+
+
+@attr.s(eq=False)
+class AwakeablePeriodicService(service.Service):
```
Comment (brettlangdon) [inline ddtrace/internal/periodic.py:319]: Should this inherit from `PeriodicService`? if I am reading this correctly the only changes are which worker class is being used and exposing `awake()`?

Comment (brettlangdon) [reply ddtrace/internal/periodic.py:319]: Do we also need a Gevent variant as well?

Comment (P403n1x87) [reply ddtrace/internal/periodic.py:319]: > Should this inherit from `PeriodicService`?

I can give this another try. I think I had found some problems when attempting it, but I will double check. Alternatively, we could add the behaviour to the `Periodic{Thread,Service}`.

> Do we also need a Gevent variant as well?

Does every thread in a Gevent app need to be a Gevent thread? If so, then probably we need that 😢

**Review by jd — APPROVED** [review]
LGTM except on the "awake" wording which I'm not sure is 100% correct but I'll defer to our 🇺🇸 people to validate the term.

**Review by mabdinur — APPROVED** [review]
The awakeable adjective looks fine to me

**Pull request merged**